### DISCUSSION TITLE
Playing URLS from list on a SD Card

### DISCUSSION THREAD

**Discussion by waszee** [pschatzmann/arduino-audio-tools#1735] [Q&A] [answered]
I need help with a coding issue.  I have stored my URL stations in a text file: "http://station1",\n, "htto:\\station2",\n .... "http:laststation".  That is one station in quotes per line of text.  I am able to read the SD Card and stored the stations into a String Array.  Putting the station lists on the SD card will let me customize the list for my grandkids or wife without having to reprogram.

Now I am stuck.  I have not figured out how to move String url[100] to const char * urls[] ={...}  which the AudioSource for the player seems to require.  Also instead of sending the whole list of possible stations to the player is there a way to just send a single url to play? 

 My current actual list only has 46 stations to choose from.  I  also reprogramed the player button keys so key1 moves index down -1, i key 2 -nothing, 3 up index+1, 4 up index10, key 5 sets the position to index and plays, 6 cycles the volume from 0 to 1.00 in 0.25 steps.  This let me select station list fairly quickly and allows to jump to different stations positions in the list.

The following reads the stations into a_url but fails to move to the urls[] const pointer array.
//
//read the url list from the file on the SD Card

'''
int readurls(){
  File url_list = SD.open("/URLs/URL_LIST.txt");
  String str;
  String a_url;
  int count = 0;
  while (url_list.available() && count < MAX_LINES) {
    str= url_list.readStringUntil('\n');
    //Serial.println(str);
    int delim = str.indexOf(',');

    //a_url=str.substring(0,delim);
    a_url=str.substring(3,delim-1);

    // Note: must copy the characters, cannot assign because array is not a class but a raw array
    strcpy(ustr[count], a_url.c_str());
    count++;
  }
  close(url_list);
  return(count);
}

'''

**Comment by pschatzmann** [comment, marked as answer]
The current AudioSourceURL implementation is designed to store a big amount of urls in PROGMEM w/o impacting the available RAM.

My recommendation would be to do something like this:
- create your own subclass of AudioSourceURL
- In your constructor open and read your file in order to count the records and update the max variable.
- in your destructor you close the file
- overwrite the following method with your own logic:  virtual const char* value(int pos)
    Read your file from the beginning until the requested index and return the string at the requested position.
    keep the last index and result in some variables, so that you can return them quickly when they are requested multiple times

This has the advantage that you can store a huge number of urls w/o impacting your RAM and that you don't need to mess up your sketch with any file processing logic!
 
ps. I suggest to change your file format and that you remove the unnecessary string delimiters which just complicate the processing

**Reply by waszee** [reply]
I think you must have been a professor at some time. Read the manuals and fix it using suggestions! I will try. 
The file format  resulted from copying from the working code station definitions that were in the {} as a csv list. I agree the quotes and comma are not needed if already have the \n delimiter in the file format.    

I was somewhat reluctant to try to change your code and wanted to work with it, but think the subclass approach will solve several problems.  I had fits trying to modify a const char value.  The strcpy issues a compiler error that it can not copy char * to const char * or something like that.  I thought I could just declare a space in memory and then modify within it but knew that used much more RAM than needed and I was not doing the declaration correctly.  My attempt does return the max station count.  It is the final line count value that realurls() returns.   I had thought about just retrieving the url at a given line but did not know how to feed that string to the existing AudioSourceURL. I might need help doing the subclass as not something I have done much.  I will come back if stuck but will do some reading first.

**Reply by waszee** [reply]
Took me a while to get your suggestions to work but have success.  The subclass was simple once I figured out I had to trim() my String values before returning as a character array.  Not sure why trim() was required.   In the subclass I  dropped const char *urls[] and added a String  to provide the pathname to the list with the url strings..  I  added a String array for storing the url strings in a protected variable.   This did not save memory but I had enough.  I counted strings using  an override of the  size() function.  I loaded the array as I counted.  The value function is a single line that  returns string at the position with the .c_str() conversion.   Before I added the trim() function to the read in strings, the value function returned but failed. The compiler gave no errors and output of those strings to the Serial monitor looked correct.   -- Next will work on adding an IR remote to AI Thinker and maybe a 1306 display.  I use those those with esp32s3 dev board and think I can give up the AI thinker buttons to get the 3 lines needed, 2 for i2c and 1 GPIO for IR remote.  Thanks again for the support.
